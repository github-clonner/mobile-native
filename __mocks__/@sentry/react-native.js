export const captureMessage = jest.fn();
export default {};
