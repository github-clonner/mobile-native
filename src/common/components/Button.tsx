//@ts-nocheck
import React, { Component } from 'react';
import {
  Text,
  ActivityIndicator,
  TouchableOpacity,
  TextStyle,
  ViewStyle,
  GestureResponderEvent,
} from 'react-native';

import ThemedStyles from '../../styles/ThemedStyles';

type PropsType = {
  text: string;
  loading?: boolean;
  onPress?: (ev: GestureResponderEvent) => void;
  textColor?: string;
  color?: string;
  children?: React.ReactNode;
  containerStyle?: ViewStyle | Array<ViewStyle>;
  accessibilityLabel?: string;
  textStyle?: TextStyle | Array<TextStyle>;
  disabled?: boolean;
  inverted?: boolean;
};

/**
 * Custom Button component
 */
export default class Button extends Component<PropsType> {
  /**
   * Default props
   */
  static defaultProps = {
    loading: false,
  };

  /**
   * Render
   */
  render() {
    const theme = ThemedStyles.style;

    const {
      onPress,
      textColor,
      color,
      children,
      containerStyle,
      accessibilityLabel,
      textStyle,
      disabled,
      loading,
      inverted,
      ...extraProps
    } = this.props;

    let background = ThemedStyles.getColor('primary_button');
    let mainColor = color || 'white';

    if (inverted !== undefined) {
      background = mainColor;
      mainColor = ThemedStyles.getColor('primary_button');
    }

    const style = { backgroundColor: background, borderRadius: 2 };

    const body = loading ? (
      <ActivityIndicator color={mainColor} />
    ) : (
      <Text style={[{ color: textColor || mainColor }, textStyle]}>
        {' '}
        {this.props.text}{' '}
      </Text>
    );

    const onButtonPress = loading ? undefined : onPress;

    return (
      <TouchableOpacity
        onPress={onButtonPress}
        disabled={disabled}
        accessibilityLabel={accessibilityLabel}
        style={[
          theme.rowJustifyCenter,
          theme.centered,
          theme.padding,
          theme.paddingHorizontal2x,
          style,
          containerStyle,
          { borderRadius: 36 },
        ]}
        {...extraProps}>
        {children}
        {body}
      </TouchableOpacity>
    );
  }
}
