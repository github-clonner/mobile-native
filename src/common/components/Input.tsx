//@ts-nocheck
import React, { Component } from 'react';
import { Text, View, StyleSheet, TouchableOpacity } from 'react-native';
import { ComponentsStyle } from '../../styles/Components';
import DateTimePicker from 'react-native-modal-datetime-picker';
import InfoPopup from './InfoPopup';
import PhoneValidationComponent from './PhoneValidationComponent';

import TextInput from './TextInput';
import ThemedStyles from '../../styles/ThemedStyles';

/**
 * Form input
 */
export default class Input extends Component {
  /**
   * State
   */
  state = {
    datePickerVisible: false,
    error: false,
  };

  showError = () => {
    this.setState({ error: true });
  };

  /**
   * Show date picker
   */
  showDatePicker = () => {
    this.setState({ datePickerVisible: true });
  };

  /**
   * Dismiss date picker
   */
  dismissDatePicker = () => {
    this.setState({ datePickerVisible: false });
  };

  /**
   * Confirm date picker
   */
  confirmDatePicker = (date) => {
    this.dismissDatePicker();
    this.props.onChangeText(date.toLocaleDateString());
  };

  /**
   * Text input
   */
  textInput = () => {
    const CS = ThemedStyles.style;
    return (
      <TextInput
        {...this.props}
        style={[CS.input, this.props.style]}
        placeholderTextColor="#444"
        returnKeyType={'done'}
        autoCapitalize={'none'}
        underlineColorAndroid="transparent"
        placeholder=""
      />
    );
  };

  /**
   * Phone input
   */
  phoneInput = () => {
    const CS = ThemedStyles.style;
    return (
      <PhoneValidationComponent
        style={[CS.input, this.props.style]}
        textStyle={CS.colorPrimaryText}
        onFocus={this.props.onFocus}
        onBlur={this.props.onBlur}
        TFA={this.props.TFA}
        TFAConfirmed={this.props.TFAConfirmed}
      />
    );
  };

  /**
   * Date input
   */
  dateInput = () => {
    const CS = ThemedStyles.style;
    const maxDate = new Date();
    maxDate.setFullYear(maxDate.getFullYear() - 13);
    return (
      <View>
        <TouchableOpacity
          {...this.props}
          style={[CS.input, this.props.style]}
          placeholderTextColor="#444"
          returnKeyType={'done'}
          autoCapitalize={'none'}
          underlineColorAndroid="transparent"
          placeholder=""
          onPress={this.showDatePicker}>
          <Text style={CS.colorPrimaryText}>{this.props.value}</Text>
        </TouchableOpacity>
        <DateTimePicker
          isVisible={this.state.datePickerVisible}
          onConfirm={this.confirmDatePicker}
          date={maxDate}
          maximumDate={maxDate}
          onCancel={this.dismissDatePicker}
          mode="date"
          display="spinner"
        />
      </View>
    );
  };

  /**
   * renders correct input
   */
  renderInput = () => {
    const inputType = this.props.inputType;
    if (inputType) {
      switch (inputType) {
        case 'textInput':
          return this.textInput();
        case 'phoneInput':
          return this.phoneInput();
        case 'dateInput':
          return this.dateInput();
      }
    }
    return this.textInput();
  };

  /**
   * Render
   */
  render() {
    const CS = ThemedStyles.style;
    const optional = this.props.optional ? (
      <Text style={[styles.optional]}>{'Optional'}</Text>
    ) : null;

    return (
      <View style={CS.marginBottom2x}>
        <View style={[styles.row]}>
          <View style={styles.row}>
            <Text style={[styles.label, this.props.labelStyle]}>
              {this.props.placeholder}
            </Text>
            {this.props.info && <InfoPopup info={this.props.info} />}
            {this.props.onError && this.state.error && (
              <View style={styles.errorContainer}>
                <Text style={[CS.colorAlert, CS.fontL, CS.textRight]}>
                  {this.props.onError}
                </Text>
              </View>
            )}
          </View>
          {optional}
        </View>
        {this.renderInput()}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  row: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  label: {
    color: '#AEB0B8',
    fontSize: 16,
    fontWeight: '600',
    marginBottom: 5,
    fontFamily: 'Roboto',
  },
  errorContainer: {
    alignContent: 'flex-end',
    width: '65%',
  },
  optional: {
    color: '#AEB0B8',
    fontSize: 14,
    fontFamily: 'Roboto-Italic',
  },
  shadow: {
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 1,
    },
    shadowOpacity: 0.1,
    shadowRadius: 1,
    elevation: 1,
  },
});
