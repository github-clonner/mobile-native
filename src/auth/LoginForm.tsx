//@ts-nocheck
import React, { Component } from 'react';

import * as Animatable from 'react-native-animatable';

import {
  View,
  Text,
  LayoutAnimation,
  // TextInput,
} from 'react-native';

import Icon from 'react-native-vector-icons/Ionicons';

import authService from './AuthService';

import i18n from '../common/services/i18n.service';
import logService from '../common/services/log.service';
import Input from '../common/components/Input';
import Button from '../common/components/Button';
import ThemedStyles from '../styles/ThemedStyles';

type PropsType = {
  onLogin: Function;
  onForgot: Function;
};

/**
 * Login Form
 */
export default class LoginForm extends Component<PropsType> {
  /**
   * State
   */
  state = {
    username: '',
    password: '',
    msg: '',
    twoFactorToken: '',
    twoFactorCode: '',
    hidePassword: true,
    inProgress: false,
    showLanguages: false,
  };

  /**
   * Constructor
   */
  constructor(props: PropsType) {
    super(props);
    this.state.language = i18n.getCurrentLocale();
  }

  /**
   * Render
   */
  render() {
    const CS = ThemedStyles.style;

    const msg = this.state.msg ? (
      <Animatable.Text
        animation="bounceInLeft"
        useNativeDriver
        style={[CS.subTitleText, CS.colorSecondaryText, CS.textCenter]}
        testID="loginMsg">
        {this.state.msg}
      </Animatable.Text>
    ) : null;

    return (
      <View style={[CS.flexContainer, CS.marginTop6x]}>
        <Text style={[CS.titleText, CS.colorPrimaryText, CS.marginBottom2x]}>
          {i18n.t('auth.login')}
        </Text>
        {msg}
        <Input
          placeholder={i18n.t('auth.username')}
          onChangeText={this.setUsername}
          value={this.state.username}
          style={CS.marginBottom2x}
          testID="usernameInput"
        />
        <View>
          <Input
            placeholder={i18n.t('auth.password')}
            secureTextEntry={this.state.hidePassword}
            onChangeText={this.setPassword}
            value={this.state.password}
            style={CS.marginBottom2x}
            testID="userPasswordInput"
          />
          <Icon
            name={this.state.hidePassword ? 'md-eye' : 'md-eye-off'}
            size={25}
            onPress={this.toggleHidePassword}
            style={CS.inputIcon}
          />
        </View>
        <Button
          onPress={() => this.onLoginPress()}
          text={i18n.t('auth.login')}
          containerStyle={[CS.button, CS.fullWidth]}
          textStyle={CS.buttonText}
          key={1}
          loading={this.state.inProgress}
          loadingRight={true}
          disabled={this.state.inProgress}
          disabledStyle={CS.backgroundTransparent}
          testID="loginButton"
        />
        <View style={CS.marginTop4x}>
          <Text style={[CS.link, CS.fontL]} onPress={this.onForgotPress}>
            {i18n.t('auth.forgot')}
          </Text>
        </View>
      </View>
    );
  }

  /**
   * Show languages
   */
  showLanguages = () => {
    this.setState({ showLanguages: true });
  };

  /**
   * Language selected
   */
  languageSelected = (language) => {
    this.setState({ language, showLanguages: false });
    i18n.setLocale(language);
  };

  /**
   * Cancel language selection
   */
  cancel = () => {
    this.setState({ showLanguages: false });
  };

  /**
   * Set two factor
   * @param {string} value
   */
  setTwoFactor = (value) => {
    const twoFactorCode = String(value).trim();
    this.setState({ twoFactorCode });
  };

  /**
   * Set two factor
   * @param {string} value
   */
  setUsername = (value) => {
    const username = String(value).trim();
    this.setState({ username });
  };

  /**
   * Set two factor
   * @param {string} value
   */
  setPassword = (value) => {
    const password = String(value).trim();
    this.setState({ password });
  };

  /**
   * Set two factor
   * @param {string} value
   */
  toggleHidePassword = () => {
    this.setState({ hidePassword: !this.state.hidePassword });
  };

  /**
   * Handle forgot password
   */
  onForgotPress = () => {
    this.props.onForgot();
  };

  /**
   * On login press
   */
  onLoginPress() {
    this.setState({ msg: '', inProgress: true });
    // is two factor auth
    if (this.state.twoFactorToken) {
      authService
        .twoFactorAuth(this.state.twoFactorToken, this.state.twoFactorCode)
        .then((data) => {
          this.props.onLogin();
        })
        .catch((err) => {
          logService.exception('[LoginForm]', err);
        });
    } else {
      authService
        .login(this.state.username, this.state.password)
        .then((data) => {
          this.props.onLogin();
        })
        .catch((errJson) => {
          LayoutAnimation.configureNext(LayoutAnimation.Presets.spring);
          if (
            errJson.error === 'invalid_grant' ||
            errJson.error === 'invalid_client'
          ) {
            this.setState({
              msg: i18n.t('auth.invalidGrant'),
              inProgress: false,
            });
            return;
          }

          //TODO implement on backend and edit
          if (errJson.error === 'two_factor') {
            this.setState({
              twoFactorToken: errJson.message,
              inProgress: false,
            });
            return;
          }

          this.setState({
            msg: errJson.message || 'Unknown error',
            inProgress: false,
          });

          logService.exception('[LoginForm]', errJson);
        });
    }
  }
}
