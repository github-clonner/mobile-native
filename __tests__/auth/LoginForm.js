import 'react-native';
import React from 'react';
import { Text, TouchableOpacity } from "react-native";
import { shallow } from 'enzyme';

import LoginForm from '../../src/auth/LoginForm';
import authService from '../../src/auth/AuthService';
import Input from '../../src/common/components/Input';
import Button from '../../src/common/components/Button';

jest.mock('../../src/auth/AuthService');

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

describe('LoginForm component', () => {
  beforeEach(() => {
    authService.login.mockClear();
  });

  it('should renders correctly', () => {
    const loginForm = renderer.create(
      <LoginForm />
    ).toJSON();
    expect(loginForm).toMatchSnapshot();
  });

  it('should calls onLogin when user login', async () => {

    authService.login.mockResolvedValue();

    const mockFn = jest.fn();

    const wrapper = shallow(
      <LoginForm onLogin={mockFn}/>
    );

    // simulate user input
    const render = wrapper.dive();
    render.find(Input).forEach(child => {
      child.simulate('changeText', 'data');
    });

    // press login
    await render.find(Button).at(0).simulate('press');

    // check state
    expect(wrapper.state().password).toEqual('data');
    expect(wrapper.state().username).toEqual('data');

    // expect auth service login to be called once
    expect(authService.login).toBeCalled();
    // expect onLogin to be called once
    expect(mockFn).toBeCalled();
  });
});
