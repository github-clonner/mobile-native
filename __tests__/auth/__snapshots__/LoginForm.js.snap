// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm component should renders correctly 1`] = `
<View
  style={
    Array [
      undefined,
      undefined,
    ]
  }
>
  <Text
    style={
      Array [
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    Login
  </Text>
  <View>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#AEB0B8",
                "fontFamily": "Roboto",
                "fontSize": 16,
                "fontWeight": "600",
                "marginBottom": 5,
              },
              undefined,
            ]
          }
        >
          Username
        </Text>
      </View>
    </View>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      editable={false}
      onChangeText={[Function]}
      placeholder=""
      placeholderTextColor="#444"
      rejectResponderTermination={true}
      returnKeyType="done"
      style={
        Array [
          undefined,
          undefined,
        ]
      }
      testID="usernameInput"
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View>
    <View>
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
            },
          ]
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#AEB0B8",
                  "fontFamily": "Roboto",
                  "fontSize": 16,
                  "fontWeight": "600",
                  "marginBottom": 5,
                },
                undefined,
              ]
            }
          >
            Password
          </Text>
        </View>
      </View>
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        editable={false}
        onChangeText={[Function]}
        placeholder=""
        placeholderTextColor="#444"
        rejectResponderTermination={true}
        returnKeyType="done"
        secureTextEntry={true}
        style={
          Array [
            undefined,
            undefined,
          ]
        }
        testID="userPasswordInput"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <Text
      allowFontScaling={false}
      onPress={[Function]}
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 25,
          },
          undefined,
          Object {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      ï‹©
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#5DBAC0",
        "borderRadius": 36,
        "opacity": 1,
      }
    }
    testID="loginButton"
  >
    <Text
      style={
        Array [
          Object {
            "color": "white",
          },
          undefined,
        ]
      }
    >
       
      Login
       
    </Text>
  </View>
  <View>
    <Text
      onPress={[Function]}
      style={
        Array [
          undefined,
          undefined,
        ]
      }
    >
      Forgot your password?
    </Text>
  </View>
</View>
`;
